### PR TITLE
SVCINT-2379 feat: smart snippets collapse

### PR DESCRIPTION
**Pull request by pathiery**
[SVCINT-2379](https://coveord.atlassian.net/browse/SVCINT-2379)

- Add ellipsis logic for smart snippet source
- Add alternative collapse ui for smart snippets

||Before|After|
|-|-|-|
|Collapsed|image|image|
|Expanded|image|image|

[SVCINT-2379]: https://coveord.atlassian.net/browse/SVCINT-2379?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx [hunk 1]
```diff
@@ -118,6 +136,15 @@ export class AtomicSmartSnippet implements InitializableComponent {
       getSmartSnippet: () => this.smartSnippet,
       getSnippetStyle: () => this.snippetStyle,
       getFeedbackSent: () => this.feedbackSent,
+      getSnippetMaximumHeight: this.snippetMaximumHeight
```
Comment (pathiery) [inline packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx:139]: I'll replace this for a `&&`

**Comment by johnnyCoveo** [comment]
@pathiery here's some change needed..

- When collapsed, we should show 3 lines of text, not just the title.
- font-size of the title (How to setup with Fitbit Connect, below the URL) is too big, it should be the same size as the top-title (which is the same.. How to setup with..)
- People also ask should have the same font-size as the titles too.

**Review by olamothe — CHANGES_REQUESTED** [review]

File: packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx [hunk 2]
```diff
@@ -101,7 +101,25 @@ export class AtomicSmartSnippet implements InitializableComponent {
 
   @State() feedbackSent = false;
 
+  @Prop({reflect: true}) public snippetMaximumHeight?: number;
+
+  @Prop({reflect: true}) public snippetCollapsedHeight?: number;
+
+  @Prop({reflect: true, attribute: 'source-url-max-length'})
+  public sourceURLMaxLength?: number;
```
[the remark this answers is not in the document]
Comment (pathiery) [reply packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx:109]: a css rule could work, but I wanted to give the option of setting the maximum length to users 🤔

File: packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx (hunk 1 above)
Comment (pathiery) [reply packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx:139]: I think I'll keep it like this, it does add 1 line here for each definition but it avoid having to deal with the case where the number is 0, thus having to make us deal with assigning `false` and everything that would entail...

File: packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (olamothe) [reply packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx:109]: As a rule of thumb, less option is generally better.

File: packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-collapse-wrapper/atomic-smart-snippet-collapse-wrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  @State() public isExpanded = true;
+
+  @State() public showButton = true;
+
+  @State() private fullHeight?: number;
+
+  private shouldRenderButton = !!this.maximumHeight;
+
+  public componentDidRender() {
+    if (this.fullHeight === undefined && this.shouldRenderButton) {
```
[the remark this answers is not in the document]
Comment (pathiery) [reply packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-collapse-wrapper/atomic-smart-snippet-collapse-wrapper.tsx:29]: I do agree with your points, but I have a few questions: 

1. Since these props are directly set in the html and are not updated by anything in the code, wouldn't changing them cause a re-render (ergo no need to watch them specifically) ? I'm not too familiar with stencil sadly....
2. `host.getBoundingClientRect().height` is indeed used to check wether we need to collapse the component or not. However, using a resize observer to check the value returned by that function would cause an infinite loop, as the value changes when collapsing the smart snippet (initial render => calculate full height => collapse => observer triggers => recalculate full height => uncollapse => recalculate => etc). This is why, like in the `atomic-smart-snippet-expendable-answer` component, the `initializeFullHeight` function is only ran once on component initialization, and not after (we just need to calculate the theoretical full height in order to asses wether the collapse logic should be activated or not). 

I'll await your advice regarding these 2 points. If listeners on the props are needed i'll add them, however I don't think the observer is necessary !

File: packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (pathiery) [reply packages/atomic/src/components/search/smart-snippets/atomic-smart-snippet/atomic-smart-snippet.tsx:109]: let me know ! cc @mikegron

File: packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-collapse-wrapper/atomic-smart-snippet-collapse-wrapper.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (olamothe) [reply packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-collapse-wrapper/atomic-smart-snippet-collapse-wrapper.tsx:29]: Ok I see. Then I guess there should be no check on `this.fullHeight === undefined` ?

[the remark this answers is not in the document]
Comment (pathiery) [reply packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-collapse-wrapper/atomic-smart-snippet-collapse-wrapper.tsx:29]: Hmmm, according to [this](https://stenciljs.com/docs/component-lifecycle#componentdidrender) didRender re-runs everytime the render function executes. In my case, on the first render, I'm checking whether the collapse feature needs to be activated, then update values accordingly, which will definitely cause a re-render. If I don't use `=== undefined` I'm once again at risk of an infinite loop which toggles the feature on and off since, as mentioned before, recalculating the high of the host will give a different value once the feature is activated. 

TLDR; in need this so that it gets run only on the first render to avoid an infinite loop !

**Review by olamothe — APPROVED** [review]

File: packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-suggestions/smart-snippet-suggestions-common.tsx [hunk 4]
```diff
@@ -197,7 +197,7 @@ export class SmartSnippetSuggestionCommon {
         <Heading
           level={this.props.getHeadingLevel()}
           part="heading"
-          class="px-6 py-4 text-2xl leading-8 border-b border-neutral"
+          class="px-6 py-4 text-xl leading-8 border-b border-neutral"
```
Comment (mikegron) [inline packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-suggestions/smart-snippet-suggestions-common.tsx:200]: Je metterais même juste large pour tous les headings?? Ca semble pas changer grand chose selon ton screenshot! @johnnyCoveo 🧑‍🌾

**Review by mikegron — APPROVED** [review]

File: packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-suggestions/smart-snippet-suggestions-common.tsx (hunk 4 above)
Comment (pathiery) [reply packages/atomic/src/components/common/smart-snippets/atomic-smart-snippet-suggestions/smart-snippet-suggestions-common.tsx:200]: @mikegron screenshot pas à jour, mais j'ai pris le xl car en px ça correspond aux figmas de Johnny !
